[Meng Qiaoran] iP

Qiaoran-M opened the pull request:
# Duke
> "The bad news is time flies. The good news is you're the pilot." -- Michael Altshuler

Are you overwhelmed by the large number of tasks to do every day? Have you ever feel worried about forgetting important tasks? Duke is here to help you remembering your tasks 🤩🤩🤩

### It's a chatbot with:
- [x] text-based user interation
- [x] auto task storage and loading 
- [x] __SUPER FAST__ and __EASY__ user experience

### By using Duke, you can:
- freely add and delete tasks
- mark task as done or undone
- list all tasks by dates and keywords

### There're 3 types of tasks provided to satisfy your need:
1. __Todo__: tasks to do ***(no time limit required)***
2. __Deadline__: task to do before end date ***(end date required)***
3. __Event__: task to do from start date to end date ***(start and end date required)***

### Duke is __FREE__ to use! 
To get started, you can directed downloaded it from [here](https://github.com/Qiaoran-M/ip).
- If you download the entire code, you can run Duke on java-supported platform by running `Duke.java`
The main function for duke startup in Duke.java in specified below for your reference:
```
public static void main(String[] args) {
        new Duke("data/Duke.txt").run();
    }
```
Take note that the task list will be auto-created and stored a text file under in data folder, and it will be auto-loaded every time you start Duke!
- If you download the .jar file, you can run Duke by typing `java -jar Duke.jar` in your local terminal

## Enjoy managing your tasks with Duke!

Qiaoran-M committed:
Add Level-1.

Qiaoran-M committed:
Add Level-2.

Qiaoran-M committed:
Add Level-3.

Qiaoran-M committed:
Add Level-3.

Qiaoran-M committed:
Add Level-4.

Qiaoran-M committed:
A-TestUiTesting

Qiaoran-M committed:
Level-5

Qiaoran-M committed:
Level-6

Qiaoran-M committed:
Add level 7.

Qiaoran-M committed:
Merge branch 'branch-Level-7'

Qiaoran-M committed:
Add level 8.

Qiaoran-M committed:
Merge branch 'branch-Level-8'

Qiaoran-M committed:
Add A-MoreOOP.

Qiaoran-M committed:
Add A-Packages.

Qiaoran-M committed:
Add A-Gradle.

Qiaoran-M committed:
Add A-JUnit.

Qiaoran-M committed:
Add A-Jar.

Qiaoran-M committed:
A-JavaDoc

Qiaoran-M committed:
Add javadoc.

Qiaoran-M committed:
Style all task classes.

Qiaoran-M committed:
Style Duke.java.

Qiaoran-M committed:
Style DukeException.java.

Qiaoran-M committed:
Style Parser.java.

Qiaoran-M committed:
Style Storage.java.

Qiaoran-M committed:
Style TaskList.java.

Qiaoran-M committed:
Style Ui.java.

Qiaoran-M committed:
Add Level 9 search by keyword method.

Qiaoran-M committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9

Qiaoran-M committed:
merge JavaDoc branch to master branch.

Qiaoran-M committed:
merge CodingStandard branch to master branch.

leonidas1712 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,41 @@
+package duke.task;
+
+import java.time.LocalDate;
+
+public class Deadline extends Task {
+    private LocalDate end;
+
+    public Deadline(String description, LocalDate end) {
+        super(description);
+        this.end = end;
+    }
```
Consider adding JavaDoc comments for this class and its public methods. I noticed that other classes in the task package also seem to lack JavaDoc comments for the necessary classes and methods, so that is something to take note of!

Qiaoran-M replied:
Thank you for pointing it out! I will add the JavaDoc comments for all task classes later.

leonidas1712 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,149 @@
+public class TaskList {
+    private List<Task> tasks;
+
+    public TaskList(List<Task> l) {
+        this.tasks = l;
+    }
+
+    public TaskList(TaskList l) {
+        this.tasks = l.tasks;
+    }
+
+    public TaskList() {
+        this.tasks = new ArrayList<Task>();
+    }
```
You can consider removing the use of `this` here (and elsewhere in the class) since the `tasks` variable is not shadowing any other variables. However, this is an 'advanced' coding standard rule so you do not need to follow it.

tantzeyeong replied:
Yes I agree with this! Might be good to standardize the style used throughout too, such as in the getTask() method, where "this" was omitted.

Qiaoran-M replied:
Thanks! I will standardize the style by remove `this`

leonidas1712 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,145 @@
+                + "\n1. todo     : tasks without any date/time attached to it"
+                + "\n2. deadline : tasks that need to be done before a specific date/time"
+                + "\n3. event    : tasks that start at a specific time and ends at a specific time");
+        System.out.println("________________________________________________________________________________________");
```
These lines are re-used throughout Ui.java for displaying output. Should you perhaps extract it to a static constant so the formatting can be kept consistent without copying and pasting the string?

leonidas1712 commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,92 @@
+    Parser parser = new Parser();
+    TaskList taskList = new TaskList();
+    @Test
+    public void parseDateTest() throws DukeException {
```
I like your separation of logical units within the method here!

leonidas1712 commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,40 @@
+
+public class TaskListTest {
+    @Test
+    public void addTaskTest() {
```
Would be nice if logical units of code were separated in this method, just like in ParserTest.java

leonidas1712 reviewed (commented):
LGTM in terms of coding standard aside from minor nits with comments and separation of logical units.

tantzeyeong commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,58 @@
+        this.isDone = false;
+    }
+
+    public String taskType() {
```
This is a method that does a certain action, so it might be better to name it as "getTaskType"?

tantzeyeong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,71 @@
+                l.addTask(Task.stringToTask(s.nextLine()));
+            }
+            return l;
+        } catch (Exception e) {
```
Might be better to catch individual exception types separately and handle them in each catch block?

tantzeyeong reviewed (commented):
Overall, this pretty much LGTM and looks very clean! Good job!

Qiaoran-M committed:
Solve conflicts between master branch adn branch level 9.

Qiaoran-M committed:
Add checkstyle configuration files and fix coding violations.

Qiaoran-M committed:
Merge branch 'branch-A-Checkstyle'

Sampy147 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,136 @@
+            }
+            taskList.addTask(new Deadline(description, end));
+            return String.format("deadline %d", taskList.getSize());
+        } else if (taskType.equals("event")) {
```
Seems like after each "if" or "else if" statement, there seem to be many lines of code manipulating the taskString. Perhaps can consider placing all these code into a method in the parser class(after the conditional statement), so that the code here would be more readable

Sampy147 reviewed (commented):
Overall, I think the OOP and structuring of classes is well done, however perhaps can shorten some lines of code, or put them in a method to improve readability

Sampy147 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,158 @@
+     */
+    public void showLoadingError() {
+        System.out.println("________________________________________________________________________________________");
+        System.out.println("☹ OOPS!!! Loading error occurred, new text file is created for task list storage.");
```
Perhaps some of these messages can be abstracted into a static constant as well to improve code readability?

Sampy147 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,158 @@
+        // for add, delete, mark / unmark
+        System.out.println("________________________________________________________________________________________");
+        Task task = taskList.getTask(Integer.parseInt(index) - 1);
+        if (command.equals("todo") || command.equals("event") || command.equals("deadline")) { //add
```
Nice work on the logic here! It's both intuitive and readable.

Qiaoran-M committed:
Add GUI images and FXML files.

Qiaoran-M committed:
Update duke package.

Qiaoran-M committed:
Add Main class.

Qiaoran-M committed:
Change dialog text display.

Qiaoran-M committed:
Merge branch 'branch-Level-10'

Qiaoran-M committed:
Add assertions to check code error.

Qiaoran-M committed:
Modify save task list function

The save function in Storage class are never used after the GUI is changed, duke chatbot cannot save task after every change.

Adding saving function to duke chatbot, modifying UI class and Parser class for correct reply message to user input

Let’s,
* change the position of save function to allow saving tasks based on each change.
* change some String returned in parseCommand function to protect abstraction principle
* change some String returned from UI to print correct message in chatbot

Qiaoran-M committed:
Modify Parser class javadoc.

Qiaoran-M committed:
Improve code quality

The code quality is not so good, since to long method is detected in the
code, and readability of the code needs improvement.

Let's shorten the length of parseCommand method in Parser class
Let's add a command package to handle different types of command
Let's add more helper methods to Parser class

Using better OOP and SLAP mechanism can make the code cleaner and more
understandable to readers.

Qiaoran-M committed:
Add tags to task.

Qiaoran-M committed:
Add list tasks by tag command.

Qiaoran-M committed:
modify test file.

Qiaoran-M committed:
Merge branch 'branch-A-Assertions'

Qiaoran-M committed:
Solve conflict between master and branch-A-CodeQuality.

Qiaoran-M committed:
Merge branch 'branch-BCD-extension'

Qiaoran-M committed:
Modify list command to handle more exceptions.

Qiaoran-M committed:
Modify bye command to better display chatbot exit Ui.

Qiaoran-M committed:
Add javadoc to new methods.

Qiaoran-M committed:
Rename chatbot and add more Javadoc to methods.

Qiaoran-M committed:
Add Javadoc to more methods and Modify GUI.

Qiaoran-M committed:
Modify GUI and upload UI picture.

Qiaoran-M committed:
Add content to README.md file.

Qiaoran-M committed:
Correct grammar mistakes in user guide.

Qiaoran-M committed:
Merge pull request #2 from Qiaoran-M/branch-A-UserGuide

Correct grammar mistakes in user guide.

Qiaoran-M committed:
Modify tests and fix bugs.

Qiaoran-M committed:
Add more error cases caused by invalid commands.

Qiaoran-M committed:
Modify UI text.
